engines/scc_proxy/lib/scc_proxy/engine.rb

```diff
@@ -24,41 +25,81 @@
   Net::HTTPRetriableError
 ].freeze
 
+INSTANCE_ID_KEYS = {
+  Amazon: 'instanceId',
+  Google: 'instance_id',
+  Microsoft: 'vmId'
+}.freeze
+
 # rubocop:disable Metrics/ModuleLength
 module SccProxy
   class << self
 
-    def headers(auth)
+    # rubocop:disable ThreadSafety/ClassAndModuleAttributes
+    attr_accessor :instance_id
+
+    # rubocop:enable ThreadSafety/ClassAndModuleAttributes
+
+    # rubocop:disable ThreadSafety/InstanceVariableInClassMethod
+    def headers(auth, params)
+      @instance_id = if params && params.class != String
+                       get_instance_id(params)
+                     else
+                       # if it is not JSON, it is the system_token already
+                       # announce system has metadata
+                       # activate product does not have metadata
+                       # so instance id comes as string
+                       params
+                     end
+
       {
         'accept' => 'application/json,application/vnd.scc.suse.com.v4+json',
         'Content-Type' => 'application/json',
-        'Authorization' => auth
+        'Authorization' => auth,
+        ApplicationController::SYSTEM_TOKEN_HEADER => @instance_id
       }
     end
+    # rubocop:enable ThreadSafety/InstanceVariableInClassMethod
+
+    def get_instance_id(params)
+      verification_provider = InstanceVerification.provider.new(
+        nil,
+        nil,
+        nil,
+        nil
+      )
```

Comment:
4 nils in a row? smells weird :D

Reply:
Yes, but if you do not include them you'd get
```bash
wrong number of arguments (given 0, expected 4)
```
Yes, we could update `engines/instance_verification/lib/instance_verification/provider_base.rb` with
```ruby
def initialize(logger=nil, request=nil, product_hash=nil, instance_data=nil)
```
But IMHO that should be a different PR, this PR aims to handle instance id as `system_token` for SCC for BYOS instances